Sources/NIO/ChannelHandlers.swift

```diff
@@ -17,6 +17,88 @@
 
 import struct Dispatch.DispatchTime
 
+
+/// `ChannelHandler` which implements some sort of backoff for a `ServerChannel` once `accept` will produce an `IOError`. This is useful as often these errors are recoverable in the sense that we can recover if we slow down the accept rate a bit.
+public final class AcceptBackoffHandler: ChannelDuplexHandler {
+    public typealias InboundIn = Channel
+    public typealias OutboundIn = Channel
+    
+    private var nextReadDeadlineNS: Int?
+    private let backoffProvider: (IOError) -> TimeAmount?
+    private var scheduledRead: Scheduled<Void>?
+    
+    /// Create a new instance
+    ///
+    /// - parameters:
+    ///     - backoffProvider: returns a `TimeAmount` to backoff more `reads` (a.k.a accepts).
+    public init(backoffProvider: @escaping (IOError) -> TimeAmount?) {
+        self.backoffProvider = backoffProvider
+    }
+    
+    public func read(ctx: ChannelHandlerContext) {
+        // If we already have a read scheduled there is no need to schedule another one.
+        guard scheduledRead == nil else { return }
+
+        if let deadline = self.nextReadDeadlineNS {
+            let now = Int(DispatchTime.now().uptimeNanoseconds)
+            if now >= deadline {
+                // The backoff already expired, just do a read.
+                doRead(ctx)
+            } else {
+                // Schedule the read to be executed after the backoff time elapsed.
+                scheduleRead(in: .nanoseconds(deadline - now), ctx: ctx)
+            }
+        } else {
+            ctx.read()
+        }
+    }
+    
+    public func errorCaught(ctx: ChannelHandlerContext, error: Error) {
+        if let ioError = error as? IOError {
+            if let amount = backoffProvider(ioError) {
+
+                self.nextReadDeadlineNS = Int(DispatchTime.now().uptimeNanoseconds) + amount.nanoseconds
+                if let scheduled = self.scheduledRead {
+                    scheduled.cancel()
+                    scheduleRead(in: amount, ctx: ctx)
+                }
```

Comment:
This code branch isn't tested, can you test it?

Reply:
done